In [ ]:
%%bash
rm -fv ipopt*
apt install -q -yy libgfortran5
sleep 1
wget -q https://www.ieor.iitb.ac.in/files/faculty/amahajan/ipopt
wget -q https://www.ieor.iitb.ac.in/files/faculty/amahajan/test.nl
chmod u+rx ipopt

removed 'ipopt'
Reading package lists...
Building dependency tree...
Reading state information...
libgfortran5 is already the newest version (8.4.0-1ubuntu1~18.04).
0 upgraded, 0 newly installed, 0 to remove and 21 not upgraded.


In [ ]:
!./ipopt test.nl



******************************************************************************
This program contains Ipopt, a library for large-scale nonlinear optimization.
 Ipopt is released as open source code under the Eclipse Public License (EPL).
         For more information visit http://projects.coin-or.org/Ipopt
******************************************************************************

This is Ipopt version 3.13.2, running with linear solver mumps.
NOTE: Other linear solvers might be more efficient (see Ipopt documentation).

Number of nonzeros in equality constraint Jacobian...:        0
Number of nonzeros in inequality constraint Jacobian.:        0
Number of nonzeros in Lagrangian Hessian.............:      230

Total number of variables............................:       30
                     variables with only lower bounds:       30
                variables with lower and upper bounds:        0
                     variables with only upper bounds:        0
Total number of equ

In [ ]:
!pip install -q pyomo

In [ ]:
import numpy as np

In [ ]:
from pyomo.environ import *

**1. ANSWER:** \
We have the following **non-linear optimization problem :** \

>$\textbf{Minimize:}$ \
\begin{align}
\bf{\sqrt{(xf-xl)^2 + (yf-yl)^2}} \\
\textbf{OR }\>\>\>\>\bf{{  (xf-xl)^2 + (yf-yl)^2}}
\end{align} 

>$\textbf{Subject to:}$
\begin{align}
\ -xf-3yf\geq \ 0 \\
\ xf-yf\geq \ 0  \\
\ xf+yf\geq \ -200 \\
\ -xf+yf\geq \ -200 \\
\ -(xl-200)^2-(yl-200)^2\geq \ -2500 \\
\text{xf, yf, xl , yl }\epsilon\textbf{ R} \\ 
\end{align}



In [ ]:
model = ConcreteModel()

In [ ]:
model.xf = Var()
model.yf = Var()
model.xl = Var()
model.yl = Var()

model.obj = Objective(expr = (model.xf-model.xl)**2 + (model.yf-model.yl)**2, sense=minimize)

model.constr1 = Constraint(expr= -model.xf - 3*model.yf >= 0)
model.constr2 = Constraint(expr= model.xf - model.yf >= 0)
model.constr3 = Constraint(expr= model.xf + model.yf >= -200)
model.constr4 = Constraint(expr= -model.xf + model.yf >= -200)
model.constr5 = Constraint(expr= -(model.xl-200)**2 - (model.yl-200)**2 >= -2500)

model.pprint()

4 Var Declarations
    xf : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    xl : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    yf : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    yl : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals

1 Objective Declarations
    obj : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : minimize : (xf - xl)**2 + (yf - yl)**2

5 Constraint Declarations
    constr1 : Size=1, Index=None, Active=True
        Key  : Lower : Body        : Upper : Active
        None :   0.0 : - xf - 3*yf :  +Inf :   True
    constr2 : Size=

In [ ]:
opt_ipopt = SolverFactory('ipopt')
result = opt_ipopt.solve(model)
print(result)

print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)
print('\nObjective = ', sqrt(model.obj()))

print('\nDecision Variables')
print('xf = ', model.xf.value)  
print('yf = ', model.yf.value)  
print('xl = ', model.xl.value) 
print('yl = ', model.yl.value) 


Problem: 
- Lower bound: -inf
  Upper bound: inf
  Number of objectives: 1
  Number of constraints: 5
  Number of variables: 4
  Sense: unknown
Solver: 
- Status: ok
  Message: Ipopt 3.13.2\x3a Optimal Solution Found
  Termination condition: optimal
  Id: 0
  Error rc: 0
  Time: 0.018199682235717773
Solution: 
- number of solutions: 0
  number of solutions displayed: 0

Solver status: ok
Solver termination condition: optimal

Objective =  202.98221256032042

Decision Variables
xf =  120.00000000096625
yf =  -39.99999999699526
xl =  184.18861162009682
yl =  152.56583486031144


**3.ANSWER :** \
location of the farm end: **(xf , yf) = (120.00000000096625, -39.99999999699526)** \
location of the lake end: **(xl , yl) = (184.18861162009682, 152.56583486031144)** \

**4. ANSWER:** \
The modified **non-linear optimization problem is :** \
>$Let \ L_{1}\ =distance \ from \ the\ lake \to the \ junction$ \
> $\> \> \> \> \ L_2 = distance \ from \ junction \to the \ farm$ \
> $\> \> \> \> \ L_3 = distance \ from \ junction \to the \ Store\ house$  

>$\textbf{Minimize:}$ \
\begin{align}
\bf{\ 0.20{\sqrt(L_3)^2 + \sqrt(L_1 + L_2)^2}} \\
\ OR \\ 
\bf{\ 0.20{(L_3)^2 + (L_1 + L_2)^2}}
\end{align} 

>$\textbf{Subject to:}$
\begin{align}
\ -xf-3yf\geq \ 0 \\
\ xf-yf\geq \ 0  \\
\ xf+yf\geq \ -200 \\
\ -xf+yf\geq \ -200 \\
\ -(xl-200)^2-(yl-200)^2\geq \ -2500 \\
\text{xf, yf, xl , yl, xj, yj }\epsilon\textbf{ R} \\ 
where \ (xj , yj) \ are \ junction \ coordinates. 
\end{align}

In [ ]:
model.xj=Var()
model.yj=Var()

In [ ]:
#deactivate the old objective
model.obj.deactivate()

In [ ]:
model.new_objective = Objective(expr =0.20*((model.xj-15)**2+(model.yj-20)**2) + (model.xj-model.xf)**2+(model.yj-model.yf)**2+(model.xj-model.xl)**2+(model.yj-model.yl)**2 ,sense = minimize)

In [ ]:
model.pprint()

6 Var Declarations
    xf : Size=1, Index=None
        Key  : Lower : Value              : Upper : Fixed : Stale : Domain
        None :  None : 120.00000000096625 :  None : False : False :  Reals
    xj : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    xl : Size=1, Index=None
        Key  : Lower : Value              : Upper : Fixed : Stale : Domain
        None :  None : 184.18861162009682 :  None : False : False :  Reals
    yf : Size=1, Index=None
        Key  : Lower : Value              : Upper : Fixed : Stale : Domain
        None :  None : -39.99999999699526 :  None : False : False :  Reals
    yj : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    yl : Size=1, Index=None
        Key  : Lower : Value              : Upper : Fixed : Stale : Domain
        None :  None : 152.5658348603114

In [ ]:
opt_ipopt = SolverFactory('ipopt')
result = opt_ipopt.solve(model)
print(result)

print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

print('\nDecision Variables')
print('xf = ', model.xf.value)  
print('yf = ', model.yf.value)  
print('xl = ', model.xl.value) 
print('yl = ', model.yl.value)
print("nJunction Coordinates:") 
print('(xj,yj)',' = ','(',model.xj.value,',', model.yj.value,')') 
 


Problem: 
- Lower bound: -inf
  Upper bound: inf
  Number of objectives: 1
  Number of constraints: 5
  Number of variables: 6
  Sense: unknown
Solver: 
- Status: ok
  Message: Ipopt 3.13.2\x3a Optimal Solution Found
  Termination condition: optimal
  Id: 0
  Error rc: 0
  Time: 0.01429605484008789
Solution: 
- number of solutions: 0
  number of solutions displayed: 0

Solver status: ok
Solver termination condition: optimal

Decision Variables
xf =  94.67459978537734
yf =  -31.55819992521143
xl =  176.51842419229084
yl =  155.85687343900707
nJunction Coordinates:
(xj,yj)  =  ( 124.6331927171219 , 58.3175788699071 )


**6.ANSWER :** \
location of the farm end: **(xf , yf) = (94.67459978537734 , -31.55819992521143)** \
location of the lake end: **(xl , yl) = (176.51842419229084 , 155.85687343900707)**